api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerInstanceTypePriceTypeAndCloudRegionTest.java

```diff
@@ -0,0 +1,428 @@
+
+        final Runnable result = () -> launchTool(configuration, INSTANCE_TYPE);
+        assertThrows(e -> e.getMessage().contains(NOT_ALLOWED), result);
+
+        verify(toolManager).loadByNameOrId(eq(IMAGE));
+        verify(instanceOfferManager).isToolInstanceAllowed(eq(INSTANCE_TYPE), eq(getResource()), eq(ID_2), eq(true));
+    }
+
+    @Test
+    public void launchPipelineShouldValidatePipelineInstanceType() {
+        launchPipeline(configuration, INSTANCE_TYPE);
+
+        verify(instanceOfferManager).isInstanceAllowed(eq(INSTANCE_TYPE), eq(ID), eq(true));
+        verify(instanceOfferManager).isPriceTypeAllowed(eq(SPOT), eq(null), eq(false));
+    }
+
+    @Test
+    public void launchPipelineShouldValidatePipelineInstanceTypeInTheSpecifiedRegion() {
+        configuration.setCloudRegionId(ID_2);
+        doReturn(NOT_ALLOWED).when(messageHelper).getMessage(eq(ERROR_INSTANCE_TYPE_IS_NOT_ALLOWED), eq(INSTANCE_TYPE));
+
+        assertThrows(e -> e.getMessage().contains(NOT_ALLOWED), () -> launchPipeline(configuration, INSTANCE_TYPE));
+
+        verify(instanceOfferManager).isInstanceAllowed(eq(INSTANCE_TYPE), eq(ID_2), eq(true));
+    }
+
+    @Test
+    public void launchPipelineShouldFailOnNotAllowedInstanceType() {
+        doReturn(false).when(instanceOfferManager).isInstanceAllowed(eq(INSTANCE_TYPE), eq(ID), eq(true));
+        doReturn(NOT_ALLOWED).when(messageHelper).getMessage(eq(ERROR_INSTANCE_TYPE_IS_NOT_ALLOWED), eq(INSTANCE_TYPE));
+
+        assertThrows(e -> e.getMessage().contains(NOT_ALLOWED), () -> launchPipeline(configuration, INSTANCE_TYPE));
+
+        verify(instanceOfferManager).isInstanceAllowed(eq(INSTANCE_TYPE), eq(ID), eq(true));
+    }
+
+    @Test
+    public void launchPipelineShouldNotValidatePipelineInstanceTypeIfItIsNotSpecified() {
+        launchPipeline(configuration, null);
+
+        verify(instanceOfferManager, times(0)).isInstanceAllowed(any(), eq(ID), eq(true));
+        verify(instanceOfferManager).isPriceTypeAllowed(eq(SPOT), eq(null), eq(false));
+    }
+
+    @Test
+    public void launchPipelineShouldValidatePriceTypeAsOnDemandIfItIsNotSpecified() {
+        configuration.setIsSpot(null);
+
+        launchPipeline(configuration, INSTANCE_TYPE);
+
+        verify(instanceOfferManager).isInstanceAllowed(eq(INSTANCE_TYPE), eq(ID), eq(false));
+        verify(instanceOfferManager).isPriceTypeAllowed(eq(ON_DEMAND), eq(null), eq(false));
+    }
+
+    @Test
+    public void launchPipelineShouldFailOnNotAllowedPriceType() {
+        doReturn(false).when(instanceOfferManager).isPriceTypeAllowed(eq(SPOT), eq(null), eq(false));
+        doReturn(NOT_ALLOWED).when(messageHelper).getMessage(eq(ERROR_PRICE_TYPE_IS_NOT_ALLOWED), eq(PriceType.SPOT));
+
+        final Runnable result = () -> launchPipeline(configuration, INSTANCE_TYPE);
+        assertThrows(e -> e.getMessage().contains(NOT_ALLOWED), result);
+
+        verify(instanceOfferManager).isInstanceAllowed(eq(INSTANCE_TYPE), eq(ID), eq(true));
```

Comment:
We test price types here so let's omit non price type checks.

Reply:
Done! Omit such case in some other tests, where we don't have to verify some mocks according to test's name.